### DISCUSSION TITLE
Relative Paths with Zettlr 2.0

### DISCUSSION THREAD

**Discussion by hildebrand-institute** [Zettlr/Zettlr#2437] [Q&A] [answered]
Hi all!
With the previous versions of Zettlr the base location for relative paths was the compiling directory of the project, with Zettlr 2.0 it seems not to be the same... Did I miss anything?
Thanks
Alex

**Comment by nathanlesage** [comment, marked as answer]
… it … should be the same? At least if I didn't accidentally introduce a bug. Could you be more specific with regard to what you're attempting to do?

**Reply by hildebrand-institute** [reply]
Pointing relatively to a file from within a LaTeX template (which does not reside into the compiling directory of the project), through something like the following, gives rise to a "not found file" error: 
`../../../dir_1/dir_2/file.tex-or-sty`;
whereas before there was no problem. Now it seems that only absolute paths are accepted.
Thanks

**Reply by nathanlesage** [reply]
Is that path relative from the template or from the file you are trying to export?

**Reply by hildebrand-institute** [reply]
From the template which is not saved within the same directory of the exporting files. 

Obviously I prefer to put all the tex templates/sty-files/tex-fragments-files into specific and higher directories. Till know there was no problem in doing so, because, no matter where the referenced file was, the relative path was always set starting from the project-exporting directory.

**Reply by nathanlesage** [reply]
Hm, that's weird. So could you please confirm if you do the following:

1. Create a template somewhere, and in there is a path that is relative and specifies a file relative to the template itself
2. Modify the defaults file in the assets manager to use that template for export
3. Export some Markdown file somewhere else

Correct?

**Reply by hildebrand-institute** [reply]
Correct! If it is anyhow relevant, I'm using the AppImage Zettlr's file on Fedora... May be installing the rpm would lead to better results, but for the time being I'd like still to maintain the previous version installed.

But, should everything work the same way as before, or did you make changes to the internal logic of relative paths?
Many thanks for your work and help

**Reply by nathanlesage** [reply]
Yes, I did make a lot of changes to the internal logic. HOWEVER what I did was not interfere with Pandoc at all. So basically running pandoc manually on the command line should yield exactly the same results as running the export command, that's why I'm really confused.

It could potentially be that the AppImage (since it's an image/virtual machine, and not a boring, regular program) somehow messes with the way Pandoc can resolve paths. Did you run Pandoc on your computer itself? That could explain the different behavior.

**Reply by hildebrand-institute** [reply]
Do you mean if I have a standalone Pandoc installation, other than the one included with Zettlr (if I remember well Zettlr comes with an internal Pandoc, right?)? Yes, I have Pandoc 2.9.2.1 installed!

**Reply by nathanlesage** [reply]
That might explain it! So here's the thing: AppImages are basically a form of virtual machine, i.e. a small mini-computer inside your computer. The reason is that inside that virtual machine we can control exactly what libraries are installed etc., so they don't interfer with potentially incompatible libraries installed on your computer. However, that also includes a lot of fancy stuff which I don't know very well myself. The only thing I know is that there is some kind of barrier between such an AppImage and your computer that has to be bridged somehow. So whenever something inside the AppImage folder wants to access anything _outside_ their container, that has to be forwarded somehow. And that is where things might go awry (although that is just a hypothesis).

Zettlr will write the defaults files to (I think) `~/.config/zettlr/defaults`, which is on your computer. On export, Zettlr will tell Pandoc (inside the AppImage) to use the corresponding defaults file, which is outside of the AppImage. It can be that basically the file is being transmitted into the container, but the relative path is still the same. However, due to a different file system, that path will suddenly point to nowhere, and as soon as Pandoc encounters it and tries to follow it, it discovers that there is no such file.

Long story short: In order to check if there's some mistake on my part, try to locate the correct defaults file (`export.pdf.yaml`) and make a copy. Then open it and in there add the input filename to `input-files` and the wanted output file to `output-file` inside that file. Then, run Pandoc and only provide the defaults file to Pandoc instead of the input- and output (`pandoc --defaults "path/to/export.pdf.yaml"`). If then the template works correctly and resolves the path, we have the culprit. (For more on how `input-files` work, see the [manual](https://pandoc.org/MANUAL.html#default-files).)

**Reply by nathanlesage** [reply]
Short addendum: I just found this note inside the code that runs Pandoc, so there does seem to be a way for it to fail regarding paths:

![image](https://user-images.githubusercontent.com/17251683/132981391-c79d551b-e521-4959-9584-c68b3215ad16.png)

**Reply by hildebrand-institute** [reply]
Leider everything seems again to be working only through absolute paths... I think, anyway, that in the next days I'll cut the Gordian knot by installing the RPM...
Vielen Dank
Alex

**Reply by hildebrand-institute** [reply]
Hi Hendrik,
I have installed Zettlr 2.0 with the rpm file on Fedora, but the issue concerning relative paths doesn't seem to be solved. The following is only an example. I'm not able to find the base directory starting from which a relative path should be set: it is not the project directory, it is not the directory where the tex template is stored... It seems therefor to be a general bug.

This is my actual directories hierarchy:
- style-files-dir
- tex-templates-dir
- doc-dir
		- imgs
		- publication
		- tex-fragments
					
The LaTeX template referenced with an absolute path from within the "defaults file" is stored into the tex-template-dir. The "publication" directory is where all the markdown files to be passed to LaTeX through Pandoc are stored (compiling dir). So if I have e.g. an image that is called from within one of the markdown files, I use the following relative path inside the markdown file: 
`../imgs/image-file-name`

But an "Unable to load picture" error is raised...

**Reply by nathanlesage** [reply]
That is indeed weird and should not happen. I can't promise a "when", but I'll keep this on my radar to see what is actually happening there.

**Reply by nathanlesage** [reply]
@hildebrand-institute Hey, I just had a look in the Pandoc documentation and found this extension: https://pandoc.org/MANUAL.html#extension-rebase_relative_paths

Would you mind giving it a try by changing the `reader` in your defaults file (the one which you want to export to) and change `markdown` to `markdown+rebase_relative_paths` and see if that works?

It seems that indeed the path resolution logic of Pandoc is a little bit more tricky than expected

**Reply by hildebrand-institute** [reply]
Hi Hendrik,
unfortunately it does not work at all. Not even if an image file is stored within the same directory of the markdown calling file... Only absolute paths work. This is very annoying for obvious reasons and represents a strong motive for downgrade to a previous Zettlr version... 

Is this so only on linux? Does nobody else use relative paths?

Thanks, anyway

**Reply by nathanlesage** [reply]
Yeah I'm using relative paths, but it works for me. Unfortunately, I cannot reproduce this bug.

**Reply by hildebrand-institute** [reply]
Well, I've also tried to install Zettlr 2.0 deb file on Debian 10, but the result is the same, only absolute paths seem to be accepted. So I'm a little bit confused, as it appears to be unrealistic that nobody else uses Debian or that I'm the only one to have the same problem both on Debian and on Fedora...

**Reply by nathanlesage** [reply]
What could also be are some faulty settings in the defaults file (see the assets manager). I tried to add many many settings and set them to reasonable defaults because I like to make it a no-brainer for new people to just export something well-looking, but I had to learn the hard way that sometimes setting anything for Pandoc actually breaks things. I have yet to find out where some of the last culprits are, and I'm almost certain that the guilty setting here is `resource-path` which I set to an empty list in the assets manager defaults. It might be that, when removed, or changed to contain only `.` (the "dot dir") everything suddenly works.

**Reply by hildebrand-institute** [reply]
Within my `export.pdf.yaml` I cannot find any `resource-path `. If I insert it, there seem to be no way to follow what indicated within the pandoc manual for `--resource-path=SEARCHPATH`... If I write: `resource-path:.`, I get "Implicit map keys need to be followed by map values" and it is the same with whatever alternative path I could imagine to write or whatever alternative syntax I use...

**Reply by nathanlesage** [reply]
I think the correct syntax is `resource-path: ["."]`

**Reply by hildebrand-institute** [reply]
It seems to be better, finally. If I use `resource-path: ["absolute_path_to_bib_files_directory"]`, then a relative path to my bib file within the yaml front matter of each paper works. 

Alternative possible syntax:
```
resource-path: 
- "absolute_path_to_bib_files_directory"
```

Still not possible, however, to use like before relative paths also within the LaTeX model. 
Vielen Dank

**Reply by nathanlesage** [reply]
I just had another look at the exporter module and I don't pass the working directory. I'm pretty sure that that might be the cause. I'll fix that and then we can see.

**Reply by nathanlesage** [reply]
So, I just changed that, and it seems to work. In half an hour or so, the next automated build will create the next nightly, so in about one and a half hour, you can download the newest nightly at https://nightly.zettlr.com (look at the modification time to make sure you don't download last week's nightly).

Please give it a try and see if that fixes the problem.

**Reply by hildebrand-institute** [reply]
YES! Everything works like a charm now.
Noch einmal vielen Dank
Alex

**Reply by nathanlesage** [reply]
Heureka!!
